pylinalg/obj/transform.py

```diff
@@ -0,0 +1,154 @@
+            self._orientation = np.array((0, 0, 0, 1))
+
+        if scale is None:
+            self._scale = np.ones(3)
+
+    def as_matrix(self):
+        """
+        The affine transforms matrix
+
+        The returned matrix is a transformation matrix that converts
+        points in homogeneous coordinates according to the current parameters
+        of this AffineTransform object. Typically, this maps from an object's
+        child frame into an object's parent frame.
+        """
+
+        result = np.zeros((4, 4))
+        result[-1, -1] = 1  # affine matrix :)
+
+        # credit to: http://www.songho.ca/opengl/gl_quaternion.htm
+        # @almarklein: Can we refactor quaternion_to_matrix to something like
+        # this? It's easier to read than the current implementation.
+        x, y, z, w = self._orientation
+        # fmt: off
+        result[:3, :3] = np.array([
+            [1 - 2*y**2 - 2*z**2,       2*x*y - 2*w*z,       2*x*z + 2*w*y],  # noqa: E201, E501
+            [      2*x*y + 2*w*z, 1 - 2*x**2 - 2*z**2,       2*y*w - 2*w*x],  # noqa: E201, E501
+            [      2*x*w - 2*w*y,       2*y*z + 2*w*x, 1 - 2*x**2 - 2*y**2],  # noqa: E201, E501
+        ]).T
+        # fmt: on
+
+        result[:, :3] = result[:, :3] * self._scale.reshape(1, 3)
+
+        # we right-multiply transformations to allow broadcasting across
+        # vectors. Hence we track the position in the bottom row
+        result[-1, :3] = self._position
+
+        return result
+
+    def inverse(self):
+        """
+        Invert this affine transformation.
+
+
+        Note
+        ----
+        The problem with this implementation is that this transform and its
+        inverse don't remain in sync, i.e., if we store the returned transform
+        somewhere and this transform updates later, the stored transform does no
+        longer invert this transform. -- We could work around this by
+        introducing `Transform.inverse_transform` and a
+        `InverseTransform(Transform)` object if desired.
+        """
+
+        scale = 1 / self.scale
+        orientation = qt.quaternion_inverse(self.orientation)
+        position = (
+            -(scale * self.position) @ qt.quaternion_to_matrix(orientation)[:3, :3]
+        )
+
+        return AffineTransform(position=position, orientation=orientation, scale=scale)
+
+    def __array__(self, dtype=None):
+        return self.transformation_matrix.astype(dtype)
+
+    def __matmul__(self, other):
+        if not isinstance(other, AffineTransform):
+            raise NotImplementedError()
+
+        other_rotation = qt.quaternion_to_matrix(other.orientation)[:3, :3]
+        position = other.position + (other.scale * self.position) @ other_rotation
+
+        orientation = qt.quaternion_multiply(self._orientation, other._orientation)
+
+        scale = self.scale * other.scale
+
+        return AffineTransform(position=position, orientation=orientation, scale=scale)
```

Comment:
I feel like this implementation (because it's not actually multiplying matrices) does not maintain the order of transformations. Is that right?

Reply:
It does maintain order :) (or at least it should ... if not there is a bug that needs fixing)

If you prefer, I could implement this as:

```python

new_transform = self.transformation_matrix @ other.transformation_matrix
position, orientation, scale = matrix_decompose(new_transform)
return AffineTransform(position=position, orientation=orientation, scale=scale)
```

The reason I didn't is that it is less accurate than working out the new `scale` and `orientation` manually. For the position, it's about the same. 

Also, we want to return a `AffineTransform` object to allow chaining statements like: `light_bulb.transform @ lamp.transform @ desk.transform @ floor.transform @ world.transform` to preserve the ability to modify/inspect the position, orientation, and scale, and to keep the possibility to optimize how we perform this computation in the future (to be more accurate, or faster, ...).

Reply:
I'm clearly not on your level math-wise, so I will just take it from you that order is maintained and accuracy is better this way. Cool!

Reply:
I'm happy to add more context. 

If we start with a classic affine matrix that includes rotation, scale, and translation then it would look something like this:

```python
scaleX*rot scaleY*rot scaleZ*rot posX
scaleX*rot scaleY*rot scaleZ*rot posY
scaleX*rot scaleY*rot scaleZ*rot posZ
0          0          0          1
```

We can segment the above into three separate matrices (`pos @ rot @ scale`). One that does the rotating, one that does translating, and one that does scaling:

```python
# scaling
scaleX 0      0      0
0      scaleY 0      0
0      0      scaleZ 0
0      0      0      1

# rotation
rot rot rot 0
rot rot rot 0
rot rot rot 0
0   0   0   1

# translation
1 0 0 posX
0 1 0 posY
0 0 1 posZ
0 0 0 1
```

The standard order in which we apply these matrices when going from child to parent is `child -> scale -> rot -> pos -> parent`. When chaining two affine transforms we simply build a longer sequence:

```
child -> scaleA -> rotA -> posA -> scaleB -> rotB -> posB -> parent
```

The million-dollar question now is: Can we reorder the terms in this sequence to bring this back to a familiar form? The answer is, ofc yes. If we respect the following rules:

```
rot -> scale <> scale -> rot  # order of scale and rotation doesn't matter
pos -> rot <> rot -> (rot @ pos)  # if I want to rotate first, I must translate with the rotated translation
pos -> scale <> scale -> (scale * pos)  # if I want to scale first, I must translate with the scaled translation

scale -> scale <> scale + scale  # two scales is the same as scaling once by the sum of scales
rot -> rot <> rot @ rot  # two rotations is the same as rotating by the combined rotation
pos -> pos <> pos + pos  # two translations is the same as translating by the sum
```

with that it becomes an exercise in simplification: all `scale`s to the left and all `pos` to the right:

```
child -> scaleA -> rotA -> posA -> scaleB -> rotB -> posB -> parent
child -> scaleA -> rotA -> scaleB -> (scaleB * posA) -> rotB -> posB -> parent
child -> scaleA -> scaleB -> rotA -> (scaleB * posA) -> rotB -> posB -> parent
child -> scaleA -> scaleB -> rotA -> rotB -> rotB @ (scaleB * posA) -> posB -> parent
child -> scaleA + scaleB -> rotA @ rotB -> rotB @ (scaleB * posA) + posB -> parent
```

This gives us our new scale, orientation, and position for `AffineMatrixA @ AffineMatrixB`:

```python
scale = scaleA + scaleB
orientation = rotA @ rot B = quaternion_multiply(orientationA, orientationB)
position = rotB @ (scaleB * positionA) + posB
```

Python calls this `self @ other`, so all that is left is substituting names.

Obviously, implementing just the last bit is faster than blowing up to matrices, multiplying, and then decomposing. What's more interesting, however, is that we also get higher numerical accuracy. If we track how each variable influences the result, it's easy to see that the new `scale` is unaffected by errors in `orientation` or position. Similarly, the new `orientation` is unaffected by any errors in `position` or `scale`. Only `position` "suffers" because we need to transform some of it as a result of the chaining.

Reply:
I see. Thanks for elaborating, cool!